File: backend/app/services/audit_storage.py
"""
app/services/audit_storage.py
Persists audit job state and results to .audit_runs/{job_id}/ directory.
State is stored as JSON, results as NDJSON (one JSON object per line).
"""
import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

RUNS_DIR = Path(".audit_runs")


def _job_dir(job_id: str) -> Path:
    return RUNS_DIR / job_id


def state_path(job_id: str) -> Path:
    return _job_dir(job_id) / "state.json"


def ndjson_path(job_id: str) -> Path:
    return _job_dir(job_id) / "results.ndjson"


def xlsx_path(job_id: str) -> Path:
    return _job_dir(job_id) / "audit.xlsx"


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def init_job_files(job_id: str) -> None:
    """Create job directory and empty results file."""
    d = _job_dir(job_id)
    d.mkdir(parents=True, exist_ok=True)
    ndjson_path(job_id).touch()


def write_state(job_id: str, state: dict) -> None:
    """Atomic write of state.json via a temp file."""
    p = state_path(job_id)
    tmp = p.with_suffix(".tmp")
    tmp.write_text(json.dumps(state, ensure_ascii=False, default=str), encoding="utf-8")
    os.replace(tmp, p)


def read_state(job_id: str) -> Optional[dict]:
    """Read state.json, returns None if missing or corrupt."""
    p = state_path(job_id)
    try:
        return json.loads(p.read_text(encoding="utf-8"))
    except Exception:
        return None


def append_ndjson(job_id: str, obj: dict) -> None:
    """Append one JSON object as a line to results.ndjson."""
    line = json.dumps(obj, ensure_ascii=False, default=str) + "\n"
    with open(ndjson_path(job_id), "a", encoding="utf-8") as f:
        f.write(line)


def read_ndjson_tail(job_id: str, max_lines: int = 50) -> list:
    """Read the last max_lines results from the NDJSON file."""
    p = ndjson_path(job_id)
    if not p.exists():
        return []

    # Read last ~200KB to avoid loading huge files
    try:
        size = p.stat().st_size
        offset = max(0, size - 200_000)
        with open(p, "rb") as f:
            f.seek(offset)
            raw = f.read().decode("utf-8", errors="replace")

        lines = [l.strip() for l in raw.splitlines() if l.strip()]
        tail = lines[-max_lines:]
        results = []
        for line in tail:
            try:
                results.append(json.loads(line))
            except Exception:
                continue
        return results
    except Exception:
        return []

File: frontend/src/lib/auditStorage.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";

export type AuditJobStatus = "queued" | "running" | "stopped" | "completed" | "failed";

export type AuditJobState = {
  job_id: string;
  status: AuditJobStatus;

  created_at: string;
  started_at?: string;
  finished_at?: string;
  stop_requested?: boolean;


  model: string;
  max_chats: number;

  processed: number;
  success: number;
  failed: number;

  total_estimate?: number;
  cursor_last_id?: string;

  last_error?: string;
  recent_errors: Array<{ at: string; request_id?: string; message: string }>;

  files: {
    ndjson: string;
    xlsx?: string;
  };
};

function runsRoot(): string {
  return path.join(process.cwd(), ".audit_runs");
}

export async function ensureRunsRoot() {
  await fsp.mkdir(runsRoot(), { recursive: true });
}

export function jobDir(jobId: string) {
  return path.join(runsRoot(), jobId);
}

export function statePath(jobId: string) {
  return path.join(jobDir(jobId), "state.json");
}

export function ndjsonPath(jobId: string) {
  return path.join(jobDir(jobId), "results.ndjson");
}

export function xlsxPath(jobId: string) {
  return path.join(jobDir(jobId), "audit.xlsx");
}

export async function initJobFiles(jobId: string) {
  await ensureRunsRoot();
  await fsp.mkdir(jobDir(jobId), { recursive: true });
  // Ensure ndjson exists
  const p = ndjsonPath(jobId);
  if (!fs.existsSync(p)) await fsp.writeFile(p, "", "utf8");
}

export async function writeState(jobId: string, state: AuditJobState) {
  // Ensure the job directory exists before writing
  await ensureRunsRoot();
  await fsp.mkdir(jobDir(jobId), { recursive: true });
  
  const p = statePath(jobId);
  const tmp = `${p}.tmp`;
  await fsp.writeFile(tmp, JSON.stringify(state, null, 2), "utf8");
  await fsp.rename(tmp, p);
}

export async function readState(jobId: string): Promise<AuditJobState | null> {
  try {
    const txt = await fsp.readFile(statePath(jobId), "utf8");
    return JSON.parse(txt);
  } catch {
    return null;
  }
}

export async function appendNdjson(jobId: string, obj: any) {
  await fsp.appendFile(ndjsonPath(jobId), JSON.stringify(obj) + "\n", "utf8");
}

export async function readNdjsonTail(jobId: string, maxLines: number) {
  // simple tail: read whole file if small, else read last ~200KB
  const p = ndjsonPath(jobId);
  const stat = await fsp.stat(p);
  const chunkSize = 200_000;
  const start = Math.max(0, stat.size - chunkSize);
  const fd = await fsp.open(p, "r");
  try {
    const buf = Buffer.alloc(stat.size - start);
    await fd.read(buf, 0, buf.length, start);
    const text = buf.toString("utf8");
    const lines = text.trim().split("\n").filter(Boolean);
    const tail = lines.slice(-maxLines);
    return tail.map((l) => {
      try { return JSON.parse(l); } catch { return { raw: l, parse_error: true }; }
    });
  } finally {
    await fd.close();
  }
}
